neutrinordp/xrdp-neutrinordp.c

```diff
@@ -609,6 +610,77 @@ lxrdp_set_param(struct mod *mod, const char *name, const char *value)
         {
             mod->perf_settings_values_mask |= PERF_DISABLE_CURSOR_SHADOW;
         }
+    }
+        else if (g_strcmp(name, "neutrinordp.allow_client_keyboardLayout") == 0)
+    {
+        if (g_text2bool(value))
+        {
+            if (settings->kbd_type == 0)
+            {
+                settings->kbd_type = mod->client_info.keyboard_type;
+            }
+            if (settings->kbd_subtype == 0)
+            {
+                settings->kbd_subtype = mod->client_info.keyboard_subtype;
+            }
+            if (settings->kbd_fn_keys == 0)
+            {
+                /* Define the most common number of function keys, 12. 
+                   because we can't get it from client. */
+                settings->kbd_fn_keys = 12;
+            }
+            if (settings->kbd_layout == 0)
+            {
+                settings->kbd_layout = mod->client_info.keylayout;
+                /* When the keyboard type is 0x07 (Japanese keyboard), 
+                   this function masks the upper 16 bits.*/
+                if (mod->client_info.keyboard_type == 0x07)
+                {
+                    LOG(LOG_LEVEL_DEBUG,
+                    "keyboardLayout mask for Japanese PC106/109 keyboard:"
+                    "[0x%08X]", 0x0000FFFF);
+                    settings->kbd_layout &= 0x0000FFFF;
+                }
+            }
+            LOG(LOG_LEVEL_INFO, "NeutrinoRDP proxy remote keyboard settings, "
+                                "kbd_type:[0x%02X], kbd_subtype:[0x%02X], "
+                                "kbd_fn_keys:[%02d], kbd_layout:[0x%08X]",
+                                settings->kbd_type, settings->kbd_subtype,
+                                settings->kbd_fn_keys, settings->kbd_layout);
+        }
+    }
+    /* The following options are for DEBUG and workaround.*/
+    else if (g_strcmp(name, "neutrinordp.override_keyboardLayout_masks") == 0)
+    {
+            settings->kbd_layout =
+                          mod->client_info.keylayout & strtol(value, NULL, 16);
+            LOG(LOG_LEVEL_WARNING,
+                "%s:[%s],overrided masked kbd_layout:[0x%08X]",
+                name, value, settings->kbd_layout);
+    }
+    else if (g_strcmp(name, "neutrinordp.override_kbd_type") == 0)
+    {
+            settings->kbd_type = strtod(value, NULL);
```

Comment:
Suggest `g_atoi()` rather than `strtod()` for this as it's used everywhere else. Or am I missing something? That applies to the following sections too.

Reply:
Thanks.
I'll rewrite it with g_atoi().

Reply:
I tried to use `g_atoi()`. This is a proprietary function of xrdp. 
Because it uses `atoi()` inside the function, it seems that it can only support decimal numbers.
https://github.com/neutrinolabs/xrdp/blob/e7fdb45744c2e8548545594042ae804dff9435ca/common/string_calls.c#L386-L395

I would like to make sure that both decimal and hexadecimal numbers can be written here.

Is there a problem with using 'strtod()'?
Would it be better to create and use the following function `g_strtod()` in `string_calls.c` ?
```
 int 
 g_strtod(const char *str) 
 { 
     if (str == 0) 
     { 
         return 0; 
     } 
       return strtod(str, NULL); 
 } 
```

Reply:
`strtod()` parses floating-point numbers, so I don't think it's what you want here. Callers will expect it to return a `double` which it doesn't. Also, callers may expect to be able to pass in the 2nd argument to `strtod()`.

A couple of options for you:-
1) Have `g_strtod()` return a `double` and also allow the second argument:-
   ```c
   double
   g_strtod(const char *str, char **endptr) 
   {
       if (str == 0) 
       { 
           str = "0";
       } 
       return strtod(str, endptr); 
   } 
   ```
   Call with

   ```c
    /* Use g_strtod() to allow for hexadecimal as well as decimal */
   settings->kbd_type = g_strtod(value, NULL);
   ```

2) Do something similar with `strtol()` instead. This allows for octal numbers as well.

3) Write a special `g_atoix()` in string_calls.c which parses decimal or hexadecimal integers only. Here's an example:-

   ```c
   /* As g_atoi() but allows for hexadecimal too */
   int g_atoix(const char *str)
   {
       int base = 10;
       if (str == NULL)
       {
           str = "0";
       }

       while (isspace(*str))
       {
           ++str;
       }

       if (*str == '0' && tolower(*(str + 1)) == 'x')
       {
           str += 2;
           base = 16;
       }
       return strtol(str, NULL, base);
   }
   ```

Hope that makes sense.

Reply:
Thank you, @matt335672 , for all your kind suggestions.

I think I will adopt your third suggestion.